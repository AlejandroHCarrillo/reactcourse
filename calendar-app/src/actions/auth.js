import { types } from "../types/types";

export const startLogin = (email, password) => {
    return async()=> { 
        console.log("email: ", email, "password: ", password);
    }
};