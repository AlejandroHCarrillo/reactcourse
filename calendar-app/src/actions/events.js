import { fetchSimple, fetchToken } from "../helpers/fetch";
import { types } from "../types/types";
import { processEvents } from "../helpers/processEvents"
import Swal from "sweetalert2";

export const eventSetActive = ( event ) => ({ 
    type: types.eventSetActive,  
    payload: event
});

export const eventStartAddNew = ( event ) => {

    return async (dispatch, getState) => {
        const { uid, name } = getState().auth;
        try{

            const resp = await fetchToken('events', event, 'POST');
            const body = await resp.json();
            
            if(body.ok){
                event.id = body.evento.id;
                event.user = {
                    uid: uid,
                    name
                }

                dispatch( eventAddNew( event ))
            }
        } catch (error){
            console.log(error);
        }
    }
}

const eventAddNew = ( event ) => ({ 
    type: types.eventAddNew, 
    payload: event
});

export const eventClearActiveEvent = () => ({ type: types.eventClearActiveEvent });

export const eventStartUpdate = ( event ) => {
    // console.log("updating ", event);
    return async ( dispatch ) =>{
        try {
            const resp = await fetchToken(`events/${event.id}`, event, 'PUT');
            // console.log(resp);
            const body = await resp.json();
            // console.log(body);
            
            if(body.ok){
                dispatch( eventUpdate(event) );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }
     
        } catch (error) {
            console.log(error);
        }
    }
}

const eventUpdate = ( event ) => ({ 
    type: types.eventUpdate,  
    payload: event
});

export const eventStartDelete = () => {  
    
    return async (dispatch, getState) => {

        const { activeEvent } = getState().calendar;
        // console.log(activeEvent);
        // console.log("deleting ", event);
        try{
            const resp = await fetchToken(`events/${activeEvent.id}`, {}, 'DELETE');
            const body = await resp.json();
            
            // console.log(body);
            if(body.ok){
                dispatch( eventDelete() );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error){
            console.log(error);
        }

    }
};

const eventDelete = () => ({  type: types.eventDelete });

export const setCurrentStartDate = ( date ) => ({  
    type: types.eventSetCurrentStartDate,
    payload: date
});

export const eventStartLoading = () => {
    return async (dispatch) => {

        try{

            const resp = await fetchToken('events', {}, 'GET');
            const body = await resp.json();
            // console.log(body.eventos);
            
            if(body.ok){
                const events = body.eventos;
                //const events = processEvents(body.eventos);
                // console.log("Los eventos con fechas tipo date:", events);
                dispatch( eventLoaded(events) );
            }
        } catch (error){
            console.log(error);
        }

    }

}

const eventLoaded = ( events ) => ({  
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({
    type: types.eventLogout
});