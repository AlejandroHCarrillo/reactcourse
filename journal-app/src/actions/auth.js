import Swal from "sweetalert2";
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";

export const startLoginWithEmailAndPassword = (email, password) =>{
    // Esto regresa un callback
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then( ({ user }) => {
            // console.log(user);
            dispatch( login(user.uid, user.displayName) );
        
            dispatch( finishLoading() );
        }).catch( e => {
            console.log(e);
            dispatch( finishLoading() );
            Swal.fire('Error de autentificacion', e.message, 'error' );
        });
    };
}

export const startRegisterWithEmailAndPassword = (email, password, name) =>{
    // Esto regresa un callback
    return ( dispatch ) =>{
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({ user }) => {
            await user.updateProfile( {profileName: name });
            // console.log(user);
            dispatch(
               login(user.uid, user.displayName)
            );
        }).catch( e => {
            Swal.fire('Error registrando al usario', e.message, 'error' );
            console.log(e);
        });
    } 
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
        .then( ({ user }) => {
            // console.log(user);
            dispatch(
               login(user.uid, user.displayName)
            );
        });
    };
};

export const login = (uid, displayName)=> ({
        type: types.login,
        payload: { 
            uid, 
            displayName }
    });

// forma larga del metodo anterior, el return esta implicito
// export const login = (uid, displayName)=> {
    //     return({
    //         type: types.login,
    //         payload: { 
    //             uid, 
    //             displayName }
    //     });
    
    // }

export const startLogout = () => {
    // llamamos asincronamente la fincion del sing out de firebase
    return async(dispatch) => {
        await firebase.auth().signOut();
        dispatch( logout() );
        dispatch( logoutCleaning());
    }
};

export const logout = () => ({
    type: types.logout
});

export const logoutCleaning = () => ({
    type: types.notesLogoutClean
});