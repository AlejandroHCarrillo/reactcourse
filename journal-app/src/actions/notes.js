import { db } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    // Como es una funcion asincrona regresamos un callback "() => {}"
    return async ( dispatch, getState ) => {
        // Con el getState obtenemos el estado del reducer en el store.js
        const uid = getState().auth.uid; 
        console.log("uid:", uid);

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
        // console.log(docRef);

        dispatch ( activeNote(docRef.id, newNote ));
    }
};

// Esta funcion no es asincrona 
// asi que podemos usar la version abreviada o corta del return 
export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
                id,
                ...note
            }
});

export const startLoadingNotes = ( uid ) => {
    return async ( dispatch ) => {
        // Cargamos las notas
        const notes = await loadNotes( uid );

        // Ponemos las notas en el state
        dispatch( setNotes(notes) );
    };
};

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
})
