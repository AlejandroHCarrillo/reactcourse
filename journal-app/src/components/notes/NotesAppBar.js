import React from 'react'
import moment from 'moment';
import { startSavingNote, startUploading } from '../../actions/notes';
import { useDispatch, useSelector } from 'react-redux';

export const NotesAppBar = ( {date} ) => {
    const noteDate = moment(date);
    const dispatch = useDispatch()
    const { active } = useSelector(state => state.notes);

    const handleSave = () => {
        dispatch(startSavingNote(active));
    }

    const handlePictureClick = () => {
        document.querySelector('#fileSelector').click();
        
    }

    const handleFileChange = (e) => {

        const file = e.target.files[0];

        if ( file ){
            dispatch( startUploading(file) );
        }

    }

    return (
        <div className="notes__appbar">
            <span>{ noteDate.format('LLLL') }</span>

            <input type="file" 
                id="fileSelector"
                name="file"
                style={{display:'none'}}
                onChange = { handleFileChange }
            ></input>

            <div>
                <button className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>

                <button className="btn"
                    onClick={ handleSave }
                >
                    Save
                </button>
            </div>
        </div>
    )
}
