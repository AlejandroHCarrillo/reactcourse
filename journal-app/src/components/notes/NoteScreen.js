import React from 'react'
import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux'
import Swal from 'sweetalert2';

import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {
    const dispatch = useDispatch();

    const { uid } = useSelector(state => state.auth );
    // console.log("El id del usuario es: ", uid );
    const { active: note } = useSelector(state => state.notes );
    const [formValues, handleInputChange, reset] = useForm(note);
    const { title, body, date } = formValues;

    // useRef permite almacenar una variable mutable
    // que no redibuja el componente si cambia
    const activeId = useRef( note.id );

    // Con este efecto reseteamos el contenido de NoteScreen
    // con el la nueva note. Usamos activeId, para validar el cambio. 
    useEffect(() => {
        if( note.id !== activeId.current ){
            reset(note);
            activeId.current = note.id;
        }
    }, [ note, reset ]);

    // Con este efecto buscamos actualizar la nota activa 
    // activeNote en el store para que este actualizada
    // usamos formValues para que lo actualize 
    // cuando cambie cualquier valor del formulario 
    useEffect(() => {
        // dispatch( activeNote(formValues.uid, { ...formValues } ) );
        dispatch( activeNote(uid, { ...formValues } ) );
    }, [formValues, dispatch])

    const handleDelete = () => {
        // console.log("Delete note id: ", id);
        Swal.fire({
            title: "¿En verdad desea eliminar esta nota?",
            text: 'Esta nota se eliminara permanentemente, ¿Desea continuar?',
            showCancelButton: true,
            showConfirmButton: true,
            confirmButtonText: `Continuar`,
            cancelButtonText: `Cancelar`,
        }).then((result) => {
            /* Read more about isConfirmed, isDenied below */
            if (result.isConfirmed) {
                dispatch( startDeleting( note.id ) );
                Swal.fire('Eliminacion exitosa!', '', 'success')
            } else if (result.isDenied) {
                Swal.fire('La nota no se elimino', '', 'info')
            }
        });
    }

    return (
        <div className="notes__main-content">                
            <NotesAppBar date={date} />
            <div className="notes__content">

                <input 
                    type="text"
                    placeholder="Escriba el titulo aqui..."
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    onChange = { handleInputChange }
                    value={ title }
                    />

                <textarea
                    placeholder="Escriba el contenido aqui..."
                    className="notes__textarea"
                    name="body"
                    onChange = { handleInputChange }
                    value={ body }
                    ></textarea>

{               note.url && 
                <div className="notes__image">
                    <img 
                        src={ note.url }
                        alt="imagen"
                        />
                </div>
}
            </div>

            <button className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>
        </div>
    )
}
