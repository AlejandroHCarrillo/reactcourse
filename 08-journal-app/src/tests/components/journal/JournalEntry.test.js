import "@testing-library/jest-dom"
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const nota = {
    id: 10,
    title: "Titulo Nota",
    body: "Nota de pruebas",
    date: 0,
    url: 'https://dominio.com/imagen.jpg'
}

const wrapper = mount( 
        <Provider store={store}>
            <JournalEntry {...nota} /> 
        </Provider>
    );


describe('Pruebas en el componente JournalEntry', () => {
    
    test('should renderizarse el componente', () => {        
        expect(wrapper).toMatchSnapshot();        
    });

    test('should activar la nota', () => {

        wrapper.find('.journal__entry').prop('onClick')();

        // const activeNotexp = activeNote(1234, { ...nota });
        // console.log("activeNotexp: ", activeNotexp  );
        // console.log(                    { "payload": 
        //         {"body": "Nota de pruebas", 
        //         "date": 0, 
        //         "id": 10, 
        //         "title": 
        //         "Titulo Nota", 
        //         "url": "https://dominio.com/imagen.jpg"
        //     }, 
        //         "type": "[Notes] Set active note"
        //     }
        // );

        expect( store.dispatch ).toHaveBeenCalled();
        expect( store.dispatch ).toHaveBeenCalledTimes(1);
        expect( store.dispatch ).toHaveBeenLastCalledWith( activeNote(1234, { ...nota }) );
        // expect( store.dispatch ).toHaveBeenLastCalledWith(
        //     { "payload": 
        //         {
        //          "body": "Nota de pruebas", 
        //          "date": 0, 
        //          "id": 10, 
        //          "title": 
        //          "Titulo Nota", 
        //          "url": "https://dominio.com/imagen.jpg"
        //          }, 
        //         "type": "[Notes] Set active note"
        //     }
        //  );


        
        
    })
    
    
})
