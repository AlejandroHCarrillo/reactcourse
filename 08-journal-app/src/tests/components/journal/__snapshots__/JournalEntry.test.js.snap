// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente JournalEntry should renderizarse el componente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://dominio.com/imagen.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Titulo Nota
    </p>
    <p
      className="journal__entry-content"
    >
      Nota de pruebas
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
